--- a/src/route.c
+++ b/src/route.c
@@ -1056,6 +1056,16 @@ static void route_mac(node_t *source, vp
 	subnet = lookup_subnet_mac(NULL, &dest, vlan_id);
 
 	if(!subnet || !subnet->owner) {
+		/* refuse packets from others with source mac owned by us */
+		if (source != myself) {
+			mac_t src;
+			memcpy(&src, &packet->data[6], sizeof src);
+			subnet = lookup_subnet_mac(myself, &src, vlan_id);
+			if(subnet && subnet->owner == myself) {
+				logger(LOG_WARNING, "Packet looping back to us from %s (%s)!", source->name, source->hostname);
+				return;
+			}
+		}
 		route_broadcast(source, packet);
 		return;
 	}
